### PR TITLE
feat: use shared informers to watch resource

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: utils/kube/kube.go [hunk 1]
```diff
@@ -88,13 +122,52 @@ func GetClientByGroupVersionKind(cfg *rest.Config, apiVersion, kind string) (dyn
 	return dc.Resource(mapping.Resource), nil
 }
 
+var clientSetCache = cache.New(time.Hour*24, time.Hour*24)
+
+func ClientSetFromRestConfig(restConfig *rest.Config) (*kubernetes.Clientset, error) {
```
Comment (moshloop) [inline utils/kube/kube.go:127]: Shouldn't we be caching by kubeconfig and not rest config ?

File: scrapers/kubernetes/events_watch.go [hunk 2]
```diff
@@ -37,8 +38,27 @@ var (
...
+
+	// The object returned by the informers do not have kind and apiversion set
+	m["kind"] = resource.Kind
+	m["apiVersion"] = resource.ApiVersion
+
+	return &unstructured.Unstructured{Object: m}, nil
+}
+
 // WatchResources watches Kubernetes resources
 func WatchResources(ctx api.ScrapeContext, config v1.Kubernetes) error {
```
Comment (moshloop) [inline scrapers/kubernetes/events_watch.go:60]: I think this should be externalised in WatchManager - if multiple kubernetes scrapers (with the same kubeconfig) are configured they should all use the same informers - Likewise updating a kubernetes scrapeconfig shouldn't make changes to the informers (unless it is to ask for them to be created

File: utils/kube/kube.go [hunk 3]
```diff
@@ -38,6 +40,18 @@ import (
 	"k8s.io/client-go/util/homedir"
 )
 
+var kubeClientCreatedCount = prometheus.NewCounterVec(
```
Comment (moshloop) [inline utils/kube/kube.go:43]: Should this be a ctx.Counter() ?

Comment (adityathebe) [reply utils/kube/kube.go:43]: kube client creation is called in a bunch of places and they don't pass in the context. So I did it this way.

File: scrapers/kubernetes/events_watch.go [hunk 4]
```diff
@@ -31,61 +31,59 @@ var (
...
 	WatchResourceBuffer = sync.Map{}
-
-	// DeleteResourceBuffer stores a buffer per kubernetes config
-	// that contains the ids of resources that have been deleted.
-	DeleteResourceBuffer = sync.Map{}
 )
 
-// WatchResources watches Kubernetes resources
+func getUnstructuredFromInformedObj(resource v1.KubernetesResourceToWatch, obj any) (*unstructured.Unstructured, error) {
+	b, err := json.Marshal(obj)
```
Comment (moshloop) [inline scrapers/kubernetes/events_watch.go:37]: Why do we need to marshal and unmarshall - It should be unstructured already ?

File: scrapers/kubernetes/informers.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+	go func() {
+		informer.Run(stopper)
+		ctx.Logger.V(1).Infof("stopped shared informer for: %v", watchResource)
+	}()
+	return nil
+}
+
+// get returns an existing shared informer instance or creates & returns a new one.
+func (t *SharedInformerManager) get(ctx api.ScrapeContext, kubeconfig, apiVersion, kind string) (cache.SharedInformer, chan struct{}, bool) {
```
Comment (moshloop) [inline scrapers/kubernetes/informers.go:101]:
```suggestion
func (t *SharedInformerManager) getOrCreate(ctx api.ScrapeContext, kubeconfig, apiVersion, kind string) (cache.SharedInformer, chan struct{}, bool) {
```

File: scrapers/kubernetes/informers.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+	cacheKey := apiVersion + kind
+
+	if val, ok := t.cache[kubeconfig]; ok {
+		if data, ok := val[cacheKey]; ok {
+			return data.informer, data.stopper, false
+		}
+	}
+
+	factory := informers.NewSharedInformerFactory(ctx.Kubernetes(), 0)
```
Comment (moshloop) [inline scrapers/kubernetes/informers.go:113]: Is this already prepopulated with kubeconfigs client ?

File: scrapers/kubernetes/events_watch.go (hunk 4 above)
Comment (adityathebe) [reply scrapers/kubernetes/events_watch.go:37]: No, we receive concrete types like core `v1.Pod`

File: scrapers/kubernetes/informers.go (hunk 6 above)
Comment (adityathebe) [reply scrapers/kubernetes/informers.go:113]: ya we set it here https://github.com/flanksource/config-db/blob/fd1e4480187e72e388b0120022290c06947434c6/scrapers/kubernetes/events_watch.go#L134

**Pull request merged**